spock-core/src/main/java/org/spockframework/runtime/extension/builtin/TimeoutInterceptor.java

```diff
@@ -123,4 +130,94 @@ public void run() {
+    System.err.printf(
+      "[spock.lang.Timeout] Method '%s' has not stopped after timing out %1.2f seconds ago - interrupting. Next try in %1.2f seconds.\n%n",
+      methodName,
+      Duration.ofNanos(now - timeoutAt).toMillis() / 1000.,
+      waitMillis / 1000.
+    );
+
+    if (unsuccessfulAttempts <= configuration.maxInterruptAttemptsWithThreadDumps) {
+      logThreadDumpOfCurrentJvm();
+      configuration.interruptAttemptListeners.forEach(Runnable::run);
+
+      if (unsuccessfulAttempts == configuration.maxInterruptAttemptsWithThreadDumps) {
+        System.out.println("[spock.lang.Timeout] No further thread dumps will be logged and no timeout listeners will be run, as the number of unsuccessful interrupt attempts exceeds configured maximum of logged attempts");
+      }
+    }
+  }
+
+  private void logMethodTimeout(String methodName, double timeoutSeconds) {
+    System.err.printf(
+      "[spock.lang.Timeout] Method '%s' timed out after %1.2f seconds.%n",
+      methodName,
+      timeoutSeconds
+    );
+
+    configuration.interruptAttemptListeners.forEach(Runnable::run);
+  }
+
+  private void logThreadDumpOfCurrentJvm() {
+    if (configuration.printThreadDumpsOnInterruptAttempts) {
+      StringBuilder sb = new StringBuilder();
+      try {
+        threadDumpCollector.appendThreadDumpOfCurrentJvm(sb);
+        System.err.println(removeThisThread(sb.toString()));
+      } catch (Throwable e) {
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        e.printStackTrace(new PrintStream(stream));
+
+        String result = "Error in attempt to fetch thread dumps: " + stream;
+        if (sb.length() > 0) {
+          result += "\n\nPartial thread dumps:\n" + sb;
+        }
+        System.err.println(result);
+      }
+    }
+  }
+
+  private static String removeThisThread(String jcmdThreadDumpOutput) {
+    Thread thisThread = Thread.currentThread();
+    String threadName = thisThread.getName();
+    long threadId = thisThread.getId();
+
+    List<String> lines = Arrays.asList(jcmdThreadDumpOutput.split("\n"));
+    Pair<Integer, Integer> thisThreadSection = findThreadSection(lines, threadName, threadId);
+    if (thisThreadSection == null) {
+      return jcmdThreadDumpOutput;
+    }
+
+    String start = TextUtil.join("\n", lines.subList(0, thisThreadSection.first()));
+    int thisThreadSectionStop = thisThreadSection.second();
+    if (thisThreadSectionStop == lines.size()) {
+      return start;
+    } else {
+      return start + TextUtil.join("\n", lines.subList(thisThreadSectionStop + 1, lines.size() - 1));
+    }
+  }
+
+  @Nullable
+  private static Pair<Integer, Integer> findThreadSection(List<String> lines, String threadName, long threadId) {
+    int threadSectionStart = -1;
+    for (int i = 0; i < lines.size(); ++i) {
+      if (lines.get(i).startsWith(String.format("\"%s\" #%d", threadName, threadId))) {
```

Comment:
🧐 extract format string as it is constant for the loop

Reply:
Done in https://github.com/spockframework/spock/pull/1658/commits/b71e5c852fcc390ed3e2ca32f5dd5de173b39150